File: js/nova/math/NumericOperation.js
var NumericOperation = function () {
	this.leftOperand = nova_null;
	this.rightOperand = nova_null;
	this.operator = nova_null;
	
	
};

NumericOperation.prototype.NumericOperation = function (operation) {
	this.this(operation);
};

NumericOperation.prototype.NumericOperation = function (left, operator, right) {
	this.this(left, operator, right);
};

NumericOperation.prototype.NumericOperation = function (left, operator, right) {
	this.this(left, operator, right);
};

NumericOperation.prototype.destroy = function () {
};

NumericOperation.prototype.this = function (operation) {
	var bounds1;
	var index1;
	var bounds2;
	var index2;
	var bounds3;
	var lh;
	var op;
	var rh;
	var type1;
	var type2;
	bounds1 = this.searchNextType(operation, 0, index1);
	bounds2 = this.searchNextType(operation, bounds1.end, index2);
	bounds3 = null;
	Console.writeLine(bounds1.toString().concat(new String(", ").concat(bounds2.toString().concat(new String(" and ").concat(Int.toString(index1).concat(new String(", ").concat(Int.toString(index2))))))));
	lh = bounds1.extractString(operation);
	op = bounds2.extractString(operation);
	rh = null;
	type1 = this.getType(lh);
	type2 = this.getType(op);
	if (type2 != this.OPERATOR) {
		if (type1 == this.OPERATOR) {
			;
		}
		rh = op;
		op = "*";
		bounds3 = bounds2.clone();
		bounds2.end = bounds2.start;
	} else {
		bounds3 = this.searchNextType(operation, bounds2.end, nova_garbageData);
		rh = bounds3.extractString(operation);
	}
	if (bounds3.end != operation.count) {
	}
};

NumericOperation.prototype.this = function (left, operator, right) {
	var leftOperator;
	var rightOperator;
	leftOperator = this.decodeOperand(left);
	rightOperator = this.decodeOperand(right);
	if (left == null) {
		;
	} else {
		;
	}
	this.this(this.leftOperand, operator, this.rightOperand);
};

NumericOperation.prototype.this = function (left, operator, right) {
	this = left;
	this = right;
	this = operator;
};

NumericOperation.prototype.decodeOperand = function (operand) {
	return null;
};

NumericOperation.prototype.getType = function (s) {
	if (s.count <= 0) {
		return 0;
	}
	return this.getType(s.chars.get(0));
};

NumericOperation.prototype.getType = function (c) {
	if (this.isNumeric(c)) {
		return this.NUMBER;
	}
	if (this.isOperator(c)) {
		return this.OPERATOR;
	}
	if (this.contains(this.INVALID_OPERATORS, c)) {
		return 0;
	}
	return this.VARIABLE;
};

NumericOperation.prototype.searchNextType = function (str, start) {
	var operators;
	var minIndex;
	var minRank;
	var index;
	var nova_local_0;
	operators = new IntArray();
	minIndex = Int.MAX_VALUE;
	minRank = Int.MIN_VALUE;
	index = this.nextOperator(str, start);
	while (index >= 0) {
		var operator;
		var rank;
		operator = str.chars.get(index);
		rank = this.getOperatorRank(operator);
		if (rank > minRank) {
			minIndex = index;
			minRank = rank;
		}
		index = this.nextOperator(str, index + 1);
	}
	if (minIndex >= str.count) {
		return null;
	}
	minIndex = this.nextNonWhitespaceIndex(str, minIndex - 1, -1) + 1;
	nova_local_0 = new Bounds(start, minIndex);
	ret1 = index;
	return nova_local_0;
};

NumericOperation.prototype.nextOperator = function (str, start) {
	var i;
	i = start;
	for (; i < str.count; i++) {
		if (this.isOperator(str.chars.get(i))) {
			return i;
		}
	}
	return -1;
};

NumericOperation.prototype.getOperatorRank = function (operator) {
	switch (operator)
	{
		case '-':
		;
		
		case '+':
		return this.ADDITION;
		
		case '/':
		;
		
		case '*':
		return this.MULTIPLICATION;
		
		case '^':
		return this.EXPONENT;
		
		case '[':
		;
		
		case '(':
			return this.GROUPING;
			
			;
		}
	};
	
	NumericOperation.prototype.nextNonWhitespaceIndex = function (str, start, direction) {
		while (start >= 0 && start < str.count && this.isWhitespace(str.chars.get(start))) {
			start = start + direction;
		}
		if (start < 0 || start >= str.count) {
			return -1;
		} else {
			return start - direction;
		}
	};
	
	NumericOperation.prototype.isLetter = function (c) {
		return !this.isOperator(c) && !this.isWhitespace(c);
	};
	
	NumericOperation.prototype.isNumeric = function (c) {
		return this.contains(this.NUMERIC_CHARS, c);
	};
	
	NumericOperation.prototype.isOperator = function (c) {
		return this.contains(this.VALID_OPERATORS, c);
	};
	
	NumericOperation.prototype.isWhitespace = function (c) {
		return this.contains(this.WHITESPACE_CHARS, c);
	};
	
	NumericOperation.prototype.contains = function (array, c) {
		var i;
		i = 0;
		for (; i < array.count; i++) {
			if (c == array.get(i)) {
				return true;
			}
		}
		return false;
	};
	
	NumericOperation.prototype.toString = function () {
		return concat(this.operator.concat(this.rightOperand.toString()));
	};
	
	NumericOperation.prototype.generated2 = function () {
		var temp;
		temp = [];
		temp = '-';
		temp = '+';
		temp = '!';
		temp = '=';
		temp = '%';
		temp = '^';
		temp = '*';
		temp = '/';
		temp = '>';
		temp = '<';
		temp = '[';
		temp = ']';
		temp = '(';
		temp = ')';
		return new CharArray(temp, 14);
	};
	
	NumericOperation.prototype.generated3 = function () {
		var temp;
		temp = [];
		temp = '`';
		temp = '~';
		temp = '@';
		temp = '#';
		temp = '$';
		temp = '&';
		temp = '_';
		temp = ';';
		temp = ':';
		temp = '\'';
		temp = '"';
		temp = '|';
		temp = '\\';
		temp = ',';
		temp = '.';
		temp = '?';
		return new CharArray(temp, 16);
	};
	
	NumericOperation.prototype.generated4 = function () {
		var temp;
		temp = [];
		temp = ' ';
		temp = '\n';
		temp = '\r';
		temp = '\t';
		return new CharArray(temp, 4);
	};
	
	NumericOperation.prototype.generated5 = function () {
		var temp;
		temp = [];
		temp = '0';
		temp = '1';
		temp = '2';
		temp = '3';
		temp = '4';
		temp = '5';
		temp = '6';
		temp = '7';
		temp = '8';
		temp = '9';
		return new CharArray(temp, 10);
	};
	
	NumericOperation.prototype.super = function () {
	};
	
	
	(function () {
			this.GROUPING = 1;
			this.EXPONENT = 2;
			this.MULTIPLICATION = 3;
			this.DIVISION = 3;
			this.ADDITION = 4;
			this.SUBTRACTION = 4;
			this.MAX = 5;
			this.INVALID = -1;
			this.OPERATOR = 1;
			this.VALID_OPERATORS = this.generated2();
			this.INVALID_OPERATORS = this.generated3();
			this.WHITESPACE_CHARS = this.generated4();
			this.NUMBER = 3;
			this.NUMERIC_CHARS = this.generated5();
			this.VARIABLE = 4;
	})();
	
		

File: java/nova/math/NumericOperation.java
package nova.math;

import java.util.Optional;
import nova.exception.ExceptionData;
import nova.exception.Exception;
import nova.exception.DivideByZeroException;
import nova.io.NovaConsole;
import nova.primitive.number.Number;
import nova.primitive.number.NovaByte;
import nova.primitive.number.NovaShort;
import nova.primitive.number.NovaInt;
import nova.primitive.number.NovaLong;
import nova.primitive.number.NovaFloat;
import nova.primitive.number.NovaDouble;
import nova.primitive.Null;
import nova.primitive.number.Char;
import nova.primitive.Bool;
import nova.datastruct.list.NovaArray;
import nova.datastruct.list.IntArray;
import nova.datastruct.list.CharArray;
import nova.datastruct.list.DoubleArray;
import nova.datastruct.list.IntRange;
import nova.thread.Thread;
import nova.thread.async.Async;
import nova.gc.GC;
import nova.math.NovaMath;
import nova.NovaObject;
import nova.NovaString;
import nova.System;
import nova.Class;
import nova.datastruct.Bounds;
import nova.math.InvalidNumericStatementException;
import nova.math.NumericOperand;

public class NumericOperation extends NumericOperand
{
	private NumericOperand leftOperand;
	private NumericOperand rightOperand;
	private NovaString operator;
	
	
	private static int GROUPING;
	private static int EXPONENT;
	private static int MULTIPLICATION;
	private static int DIVISION;
	private static int ADDITION;
	private static int SUBTRACTION;
	private static int MAX;
	private static int INVALID;
	private static int OPERATOR;
	private static CharArray VALID_OPERATORS;
	private static CharArray INVALID_OPERATORS;
	private static CharArray WHITESPACE_CHARS;
	private static int NUMBER;
	private static CharArray NUMERIC_CHARS;
	private static int VARIABLE;
	
	
	
	public NumericOperation(NovaString operation)
	{
		init(operation);
	}
	
	public NumericOperation(NovaString left, NovaString operator, NovaString right)
	{
		init(left, operator, right);
	}
	
	private NumericOperation(NumericOperand left, NovaString operator, NumericOperand right)
	{
		init(left, operator, right);
	}
	
	public void init(NovaString operation)
	{
		Bounds bounds1;
		int index1;
		Bounds bounds2;
		int index2;
		Bounds bounds3;
		NovaString lh;
		NovaString op;
		NovaString rh;
		int type1;
		int type2;
		bounds1 = searchNextType(operation, 0);
		bounds2 = searchNextType(operation, bounds1.end);
		bounds3 = null;
		NovaConsole.writeLine(bounds1.toString().concat(new NovaString(", ").concat(bounds2.toString().concat(new NovaString(" and ").concat(NovaInt.toString(index1).concat(new NovaString(", ").concat(NovaInt.toString(index2))))))));
		lh = bounds1.extractString(operation);
		op = bounds2.extractString(operation);
		rh = null;
		type1 = getType(lh);
		type2 = getType(op);
		if (type2 != OPERATOR)
		{
			if (type1 == OPERATOR)
			{
				throw new InvalidNumericStatementException(new NovaString("Missing left hand operand in operation '").concat(operation.concat(new NovaString("'"))));
			}
			rh = op;
			op = new NovaString("*");
			bounds3 = bounds2.clone();
			bounds2.end = bounds2.start;
		}
		else
		{
			bounds3 = searchNextType(operation, bounds2.end);
			rh = bounds3.extractString(operation);
		}
		if (bounds3.end != operation.count)
		{
		}
	}
	
	public void init(NovaString left, NovaString operator, NovaString right)
	{
		NumericOperand leftOperator;
		NumericOperand rightOperator;
		leftOperator = decodeOperand(left);
		rightOperator = decodeOperand(right);
		if (left == null)
		{
			throw new Exception(new NovaString("Left operand '").concat(left.concat(new NovaString("' is invalid"))));
		}
		else if (right == null)
		{
			throw new Exception(new NovaString("Right operand '").concat(right.concat(new NovaString("' is invalid"))));
		}
		init(leftOperand, operator, rightOperand);
	}
	
	public void init(NumericOperand left, NovaString operator, NumericOperand right)
	{
		this.leftOperand = left;
		this.rightOperand = right;
		this.operator = operator;
	}
	
	private NumericOperand decodeOperand(NovaString operand)
	{
		return null;
	}
	
	private static int getType(NovaString s)
	{
		if (s.count <= 0)
		{
			return 0;
		}
		return getType(s.chars.get(0));
	}
	
	private static int getType(char c)
	{
		if (isNumeric(c))
		{
			return NUMBER;
		}
		if (isOperator(c))
		{
			return OPERATOR;
		}
		if (contains(INVALID_OPERATORS, c))
		{
			return 0;
		}
		return VARIABLE;
	}
	
	private static Bounds searchNextType(NovaString str, int start)
	{
		IntArray operators;
		int minIndex;
		int minRank;
		int index;
		Bounds nova_local_0;
		operators = new IntArray();
		minIndex = NovaInt.MAX_VALUE;
		minRank = NovaInt.MIN_VALUE;
		index = nextOperator(str, start);
		while (index >= 0)
		{
			char operator;
			int rank;
			operator = str.chars.get(index);
			rank = getOperatorRank(operator);
			if (rank > minRank)
			{
				minIndex = index;
				minRank = rank;
			}
			index = nextOperator(str, index + 1);
		}
		if (minIndex >= str.count)
		{
			return null;
		}
		minIndex = nextNonWhitespaceIndex(str, minIndex - 1, -1) + 1;
		nova_local_0 = new Bounds(start, minIndex);
		ret1 = index;
		return nova_local_0;
	}
	
	private static int nextOperator(NovaString str, int start)
	{
		int i;
		i = (int)start;
		for (; i < (int)str.count; i++)
		{
			if (isOperator(str.chars.get(i)))
			{
				return i;
			}
		}
		return -1;
	}
	
	private static int getOperatorRank(char operator)
	{
		switch (operator)
		{
			case '-':
			
			case '+':
			return ADDITION;
			
			case '/':
			
			case '*':
			return MULTIPLICATION;
			
			case '^':
			return EXPONENT;
			
			case '[':
			
			case '(':
				return GROUPING;
				
				default:
				return INVALID;
				
			}
		}
		
		private static int nextNonWhitespaceIndex(NovaString str, int start, int direction)
		{
			while (start >= 0 && start < str.count && isWhitespace(str.chars.get(start)))
			{
				start = start + direction;
			}
			if (start < 0 || start >= str.count)
			{
				return -1;
			}
			else
			{
				return start - direction;
			}
		}
		
		private static boolean isLetter(char c)
		{
			return !isOperator(c) && !isWhitespace(c);
		}
		
		private static boolean isNumeric(char c)
		{
			return contains(NUMERIC_CHARS, c);
		}
		
		private static boolean isOperator(char c)
		{
			return contains(VALID_OPERATORS, c);
		}
		
		private static boolean isWhitespace(char c)
		{
			return contains(WHITESPACE_CHARS, c);
		}
		
		private static boolean contains(CharArray array, char c)
		{
			int i;
			i = (int)0;
			for (; i < (int)array.count; i++)
			{
				if (c == array.get(i))
				{
					return true;
				}
			}
			return false;
		}
		
		public NovaString toString()
		{
			return leftOperand.toString().concat(operator.concat(rightOperand.toString()));
		}
		
		private static CharArray generated2()
		{
			char[] temp;
			temp = new char[14];
			temp[0] = '-';
			temp[1] = '+';
			temp[2] = '!';
			temp[3] = '=';
			temp[4] = '%';
			temp[5] = '^';
			temp[6] = '*';
			temp[7] = '/';
			temp[8] = '>';
			temp[9] = '<';
			temp[10] = '[';
			temp[11] = ']';
			temp[12] = '(';
			temp[13] = ')';
			return new CharArray(temp, 14);
		}
		
		private static CharArray generated3()
		{
			char[] temp;
			temp = new char[16];
			temp[0] = '`';
			temp[1] = '~';
			temp[2] = '@';
			temp[3] = '#';
			temp[4] = '$';
			temp[5] = '&';
			temp[6] = '_';
			temp[7] = ';';
			temp[8] = ':';
			temp[9] = '\'';
			temp[10] = '"';
			temp[11] = '|';
			temp[12] = '\\';
			temp[13] = ',';
			temp[14] = '.';
			temp[15] = '?';
			return new CharArray(temp, 16);
		}
		
		private static CharArray generated4()
		{
			char[] temp;
			temp = new char[4];
			temp[0] = ' ';
			temp[1] = '\n';
			temp[2] = '\r';
			temp[3] = '\t';
			return new CharArray(temp, 4);
		}
		
		private static CharArray generated5()
		{
			char[] temp;
			temp = new char[10];
			temp[0] = '0';
			temp[1] = '1';
			temp[2] = '2';
			temp[3] = '3';
			temp[4] = '4';
			temp[5] = '5';
			temp[6] = '6';
			temp[7] = '7';
			temp[8] = '8';
			temp[9] = '9';
			return new CharArray(temp, 10);
		}
		
	}
		